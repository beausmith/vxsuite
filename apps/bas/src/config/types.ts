// Events
export type EventTargetFunction = (event: React.FormEvent<EventTarget>) => void;
export type ButtonEvent = React.MouseEvent<HTMLButtonElement>;
export type ButtonEventFunction = (event: ButtonEvent) => void;
