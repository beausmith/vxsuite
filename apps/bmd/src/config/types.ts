import React from 'react'

// Generic
export interface Dictionary<T> {
  [key: string]: T | undefined
}

// Events
export type InputEvent = React.FormEvent<EventTarget>
export type ButtonEvent = React.MouseEvent<HTMLButtonElement>

// Election
export interface Candidate {
  readonly id: string
  readonly name: string
  readonly party?: string
  isWriteIn?: boolean
}
export type OptionalCandidate = Candidate | undefined
export type Vote = Candidate[] // will eventually add "| Approval" to this type
export type OptionalVote = Candidate[] | undefined
export type VotesDict = Dictionary<Vote>
export interface Contest {
  readonly id: string
  readonly section?: string
  readonly title: string
  readonly type: string
}
export interface CandidateContest extends Contest {
  readonly type: 'candidate'
  readonly seats: number
  readonly candidates: Candidate[]
  readonly allowWriteIns: boolean
}
export interface YesNoContest extends Contest {
  readonly type: 'yesno'
  readonly description: string
}
export interface BMDConfig {
  readonly requireActivation?: boolean
  readonly showHelpPage?: boolean
  readonly showSettingsPage?: boolean
}
export interface ElectionDefaults {
  readonly bmdConfig: BMDConfig
}
export interface Election {
  readonly contests: Array<CandidateContest | YesNoContest>
  readonly county: string
  readonly date: string
  readonly seal: string
  readonly state: string
  readonly title: string
  readonly bmdConfig?: BMDConfig
}
export type OptionalElection = Election | undefined
export type UpdateVoteFunction = (contestId: string, vote: Vote) => void
export interface BallotContextInterface {
  readonly election: Election | undefined
  resetBallot: () => void
  setBallotKey: (activationCode: string) => void
  updateVote: UpdateVoteFunction
  votes: VotesDict
}

export default {}
