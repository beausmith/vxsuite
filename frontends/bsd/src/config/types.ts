import {
  BallotId,
  BallotStyleId,
  Dictionary,
  ElectionDefinition,
  MachineId,
  MarkInfo,
  PrecinctId,
  SerializableBallotPageLayout,
} from '@votingworks/types';
import { z } from 'zod';

export interface MachineConfig {
  machineId: string;
  bypassAuthentication: boolean;
}
export const MachineConfigSchema: z.ZodSchema<MachineConfig> = z.object({
  machineId: MachineId,
  bypassAuthentication: z.boolean(),
});

export type MachineConfigResponse = MachineConfig;
export const MachineConfigResponseSchema = MachineConfigSchema;

// Events
export type EventTargetFunction = (event: React.FormEvent<EventTarget>) => void;
export type InputEvent = React.FormEvent<EventTarget>;
export type ButtonEvent = React.MouseEvent<HTMLButtonElement>;
export type ButtonEventFunction = (event: ButtonEvent) => void;

// Election
export type SetElectionDefinition = (value?: ElectionDefinition) => void;

// Scanner Types
export interface CastVoteRecord
  extends Dictionary<string | string[] | boolean> {
  _precinctId: PrecinctId;
  _ballotStyleId: BallotStyleId;
  _ballotType: 'absentee' | 'provisional' | 'standard';
  _ballotId: BallotId;
  _testBallot: boolean;
  _scannerId: string;
}

export type Ballot = BmdBallotInfo | HmpbBallotInfo | UnreadableBallotInfo;

export interface BmdBallotInfo {
  id: number;
  filename: string;
  cvr: CastVoteRecord;
}

export interface HmpbBallotInfo {
  id: number;
  filename: string;
  cvr: CastVoteRecord;
  marks: MarkInfo;
  layout: SerializableBallotPageLayout;
}

export interface UnreadableBallotInfo {
  id: number;
  filename: string;
}
