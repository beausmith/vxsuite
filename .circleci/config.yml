version: 2.1

executors:
  nodejs-browsers:
    docker:
      - image: circleci/node:12.19.0-browsers
  nodejs:
    docker:
      - image: circleci/node:12.19

jobs:
  test-ballot-encoder:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/ballot-encoder build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/ballot-encoder lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/ballot-encoder test

  test-bas:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir apps/bas build
      - run:
          name: Lint
          command: |
            pnpm --dir apps/bas lint
      - run:
          name: Test
          command: |
            pnpm --dir apps/bas test

  test-bmd-main:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir apps/bmd build
      - run:
          name: Lint
          command: |
            pnpm --dir apps/bmd lint
      - run:
          name: Test
          command: |
            pnpm --dir apps/bmd test

  test-bmd-integration:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir apps/bmd build
      - run:
          name: Lint
          command: |
            pnpm --dir apps/bmd lint
      - run:
          name: Test
          command: |
                  pnpm --dir apps/bmd test:e2e

  test-bsd:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir apps/bsd build
      - run:
          name: Lint
          command: |
            pnpm --dir apps/bsd lint
      - run:
          name: Test
          command: |
            pnpm --dir apps/bsd test

  test-election-manager:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir apps/election-manager build
      - run:
          name: Lint
          command: |
            pnpm --dir apps/election-manager lint
      - run:
          name: Test
          command: |
            pnpm --dir apps/election-manager test

  test-eslint-plugin-vx:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/eslint-plugin-vx build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/eslint-plugin-vx lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/eslint-plugin-vx test

  test-fixtures:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/fixtures build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/fixtures lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/fixtures test

  test-logging:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/logging build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/logging lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/logging test

  test-hmpb-interpreter:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/hmpb-interpreter build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/hmpb-interpreter lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/hmpb-interpreter test

  test-lsd:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/lsd build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/lsd lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/lsd test

  test-module-converter-ms-sems:
    docker:
      - image: circleci/python:3.7.7
    steps:
      - checkout
      - run:
          name: Dependencies
          command: |
            sudo apt update -y
            make -C apps/module-converter-ms-sems install-dependencies
            make -C apps/module-converter-ms-sems install-dev-dependencies
      - run:
          name: Test
          command: |
            make -C apps/module-converter-ms-sems coverage
      # - run:
      #     name: Typecheck
      #     command: |
      #       make -C apps/module-converter-ms-sems typecheck

  test-module-scan:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir apps/module-scan build
      - run:
          name: Lint
          command: |
            pnpm --dir apps/module-scan lint
      - run:
          name: Test
          command: |
            pnpm --dir apps/module-scan test

  test-module-smartcards:
    docker:
      - image: circleci/python:3.7.7
    steps:
      - checkout
      - run:
          name: Dependencies
          command: |
            sudo apt update -y
            make -C apps/module-smartcards install-smartcard
            make -C apps/module-smartcards build-dev
      - run:
          name: Test
          command: |
            make -C apps/module-smartcards coverage
      - run:
          name: Typecheck
          command: |
            make -C apps/module-smartcards typecheck

  test-plustek-sdk:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/plustek-sdk build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/plustek-sdk lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/plustek-sdk test

  test-precinct-scanner:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir apps/precinct-scanner build
      - run:
          name: Lint
          command: |
            pnpm --dir apps/precinct-scanner lint
      - run:
          name: Test
          command: |
            pnpm --dir apps/precinct-scanner test

  test-types:
    executor: nodejs
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/types build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/types lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/types test

  test-ui:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/ui build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/ui lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/ui test

  test-utils:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/utils build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/utils lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/utils test

  test-test-utils:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir libs/test-utils build
      - run:
          name: Lint
          command: |
            pnpm --dir libs/test-utils lint
      - run:
          name: Test
          command: |
            pnpm --dir libs/test-utils test

  test-codemods:
    executor: nodejs-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir codemods build
      - run:
          name: Lint
          command: |
            pnpm --dir codemods lint
      - run:
          name: Test
          command: |
            pnpm --dir codemods test

  test-integration-testing-bsd:
    docker:
      - image: circleci/python:3.7.7-node-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir integration-testing/bsd build
      - run:
          name: Lint
          command: |
            pnpm --dir integration-testing/bsd lint
      - run:
          name: Test
          command: |
            pnpm --dir integration-testing/bsd test

  test-integration-testing-election-manager:
    docker:
      - image: circleci/python:3.7.7-node-browsers
    resource_class: xlarge
    steps:
      - checkout-and-install
      - run:
          name: Build
          command: |
            pnpm --dir integration-testing/election-manager build
      - run:
          name: Lint
          command: |
            pnpm --dir integration-testing/election-manager lint
      - run:
          name: Test
          command: |
            pnpm --dir integration-testing/election-manager test

workflows:
  test:
    jobs:
      - test-ballot-encoder
      - test-bas
      - test-bmd-main
      - test-bmd-integration
      - test-bsd
      - test-election-manager
      - test-eslint-plugin-vx
      - test-fixtures
      - test-hmpb-interpreter
      - test-integration-testing-bsd
      - test-integration-testing-election-manager
      - test-logging
      - test-lsd
      - test-module-converter-ms-sems
      - test-module-scan
      - test-module-smartcards
      - test-plustek-sdk
      - test-precinct-scanner
      - test-test-utils
      - test-types
      - test-ui
      - test-utils

commands:
  checkout-and-install:
    description: Get the code and install dependencies.
    steps:
      - run:
          name: Install pnpm
          command: |
            curl -f https://get.pnpm.io/v6.js | sudo node - add --global pnpm@5
      - checkout
      # Edit this comment somehow in order to invalidate the CircleCI cache.
      # Since the contents of this file affect the cache key, editing only a
      # comment will invalidate the cache without changing the behavior.
      - restore_cache:
          key:
            dotcache-cache-{{checksum ".circleci/config.yml" }}-{{ checksum
            "pnpm-lock.yaml" }}
      - run:
          name: Setup Dependencies
          command: pnpm install --frozen-lockfile
      - save_cache:
          key:
            dotcache-cache-{{checksum ".circleci/config.yml" }}-{{ checksum
            "pnpm-lock.yaml" }}
          paths:
            - ~/.pnpm-store
