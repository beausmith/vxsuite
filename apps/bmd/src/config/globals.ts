export default {
  FONT_SIZES: [18, 24, 28, 32],
  TEXT_SIZE: 1,
}
