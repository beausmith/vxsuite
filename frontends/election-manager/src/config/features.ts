import { unsafeParse } from '@votingworks/types';
import { asBoolean } from '@votingworks/utils';
import { ConverterClientType, ConverterClientTypeSchema } from './types';

/**
 * Determines whether authentication is enabled.
 *
 * To disable authentication, add this line to `frontends/election-manager/.env.local`:
 *
 *     REACT_APP_VX_BYPASS_AUTHENTICATION=true
 *
 * To enable it, remove the line or comment it out. Restarting the server is required.
 *
 * @see https://create-react-app.dev/docs/adding-custom-environment-variables/
 */
export function isAuthenticationEnabled(): boolean {
  return (
    process.env.NODE_ENV !== 'development' ||
    !asBoolean(process.env.REACT_APP_VX_BYPASS_AUTHENTICATION)
  );
}

/**
 * Determines whether write-in adjudication is enabled.
 *
 * To enable write-in adjudication, add this line to `frontends/election-manager/.env.local`:
 *
 *     REACT_APP_VX_ENABLE_WRITE_IN_ADJUDICATION=true
 *
 * To disable it, remove the line or comment it out. Restarting the server is required.
 *
 * @see https://create-react-app.dev/docs/adding-custom-environment-variables/
 */
export function isWriteInAdjudicationEnabled(): boolean {
  return (
    // Bundlers can tell this evaluates to `false` when building for production,
    // and can inline the function call. Thus, any uses of this function may
    // also enable further tree shaking when used in a way that the bundler can
    // understand, e.g. `if (isWriteInAdjudicationEnabled())` or
    // `isWriteInAdjudicationEnabled() && ...`.
    process.env.NODE_ENV === 'development' &&
    asBoolean(process.env.REACT_APP_VX_ENABLE_WRITE_IN_ADJUDICATION)
  );
}

/**
 * Determines which converter client to use, if any.
 */
export function getConverterClientType(): ConverterClientType | undefined {
  const rawConverterClientType = process.env.REACT_APP_VX_CONVERTER;

  if (!rawConverterClientType) {
    return;
  }

  return unsafeParse(ConverterClientTypeSchema, rawConverterClientType);
}
