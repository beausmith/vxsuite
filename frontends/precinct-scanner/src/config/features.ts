import { asBoolean } from '@votingworks/utils';

/**
 * Determines whether authentication is enabled.
 *
 * To disable authentication, add this line to `frontends/bsd/.env.local`:
 *
 *     REACT_APP_VX_BYPASS_AUTHENTICATION=true
 *
 * To enable it, remove the line or comment it out. Restarting the server is required.
 *
 * @see https://create-react-app.dev/docs/adding-custom-environment-variables/
 */
export function isAuthenticationEnabled(): boolean {
  return (
    process.env.NODE_ENV !== 'development' ||
    !asBoolean(process.env.REACT_APP_VX_BYPASS_AUTHENTICATION)
  );
}
