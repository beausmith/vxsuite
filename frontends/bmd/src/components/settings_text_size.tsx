import React from 'react';
import styled from 'styled-components';

import { Button, SegmentedButton } from '@votingworks/ui';

import { SetUserSettings, UserSettings } from '../config/types';
import { SIZE_THEMES, TEXT_SIZES } from '../config/globals';

const TextSizeSegmentedButton = styled(SegmentedButton)`
  button {
    min-width: ${TEXT_SIZES[1] * 3.5}px;
    /* stylelint-disable declaration-no-important */
    &[data-size='S'] {
      font-size: ${TEXT_SIZES[0]}px !important;
    }
    &[data-size='M'] {
      font-size: ${TEXT_SIZES[1]}px !important;
    }
    &[data-size='L'] {
      font-size: ${TEXT_SIZES[2]}px !important;
    }
    &[data-size='XL'] {
      font-size: ${TEXT_SIZES[3]}px !important;
    }
    /* stylelint-enable */
  }
`;

const Label = styled.label`
  display: block;
  margin-bottom: 0.4rem;
`;

interface Props {
  userSettings: UserSettings;
  setUserSettings: SetUserSettings;
}

const ariaLabels = ['Small', 'Medium', 'Large', 'Extra Large'];

export function SettingsTextSize({
  userSettings,
  setUserSettings,
}: Props): JSX.Element {
  return (
    <p>
      <Label aria-hidden>Text Size</Label>
      <TextSizeSegmentedButton data-testid="change-text-size-buttons">
        {SIZE_THEMES.map((v, i) => (
          <Button
            key={v}
            data-size={v}
            onPress={() => setUserSettings({ sizeTheme: v })}
            value={v}
            primary={userSettings.sizeTheme === v}
            aria-label={`${
              userSettings.sizeTheme === v ? 'Selected' : ''
            } Text Size: ${ariaLabels[i]}`}
          >
            A
          </Button>
        ))}
      </TextSizeSegmentedButton>
    </p>
  );
}
