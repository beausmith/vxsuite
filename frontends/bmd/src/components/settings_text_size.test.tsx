import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { DEFAULT_USER_SETTINGS } from '../config/globals';
import { UserSettings } from '../config/types';
import { SettingsTextSize } from './settings_text_size';

const userSettings: UserSettings = DEFAULT_USER_SETTINGS;
const setUserSettings = jest.fn();

it('renders SettingsTextSize', () => {
  const { container, getAllByText } = render(
    <SettingsTextSize
      userSettings={userSettings}
      setUserSettings={setUserSettings}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
  const buttons = getAllByText('A');
  expect(buttons.length).toBe(4);
  fireEvent.click(buttons[0]);
  expect(setUserSettings).toBeCalledWith({ sizeTheme: 'S' });
  fireEvent.click(buttons[1]);
  expect(setUserSettings).toBeCalledWith({ sizeTheme: 'M' });
  fireEvent.click(buttons[2]);
  expect(setUserSettings).toBeCalledWith({ sizeTheme: 'L' });
  fireEvent.click(buttons[3]);
  expect(setUserSettings).toBeCalledWith({ sizeTheme: 'XL' });
});
