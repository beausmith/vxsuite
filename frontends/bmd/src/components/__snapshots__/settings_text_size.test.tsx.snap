// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SettingsTextSize 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
}

.c1 > button {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-shadow: inset 1px 0 0 rgb(190,190,190);
}

.c1 > button:first-child {
  box-shadow: none;
}

.c1 > button:not(:last-child) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.c1 > button:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.c1 > button:disabled {
  background: #028099;
  color: #ffffff;
}

.c1 button {
  min-width: 98px;
}

.c1 button[data-size='S'] {
  font-size: 22px !important;
}

.c1 button[data-size='M'] {
  font-size: 28px !important;
}

.c1 button[data-size='L'] {
  font-size: 36px !important;
}

.c1 button[data-size='XL'] {
  font-size: 56px !important;
}

.c0 {
  display: block;
  margin-bottom: 0.4rem;
}

<p>
  <label
    aria-hidden="true"
    class="c0"
  >
    Text Size
  </label>
  <span
    class="sc-bqiQRQ c1"
    data-testid="change-text-size-buttons"
  >
    <button
      aria-label=" Text Size: Small"
      class="sc-egiSv dHUDgX"
      data-size="S"
      type="button"
      value="S"
    >
      A
    </button>
    <button
      aria-label="Selected Text Size: Medium"
      class="sc-egiSv KWJA-d"
      data-size="M"
      type="button"
      value="M"
    >
      A
    </button>
    <button
      aria-label=" Text Size: Large"
      class="sc-egiSv dHUDgX"
      data-size="L"
      type="button"
      value="L"
    >
      A
    </button>
    <button
      aria-label=" Text Size: Extra Large"
      class="sc-egiSv dHUDgX"
      data-size="XL"
      type="button"
      value="XL"
    >
      A
    </button>
  </span>
</p>
`;
